### ISSUE TITLE
How to setup a pre-commit hook for Elixir / Phoenix projects

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-pre-commit#6]
we _all_ know how to setup a pre-commit hook for our Node.js projects ... ✅ 
but _how_ is it done for `elixir-lang` / `phoenix` projects...? 🤔

**Comment by Jbarget** [comment]
this would be 👌 

Did a little poking around, and this looks promising: https://mattvonrocketstein.github.io/heredoc/elixir-boilerplate.html

it uses this: http://pre-commit.com/

**Comment by Jbarget** [comment]
So i've got it working [here](https://github.com/dwyl/tudo/pull/161) 🎉  - but theres a difference from what we've been used to with the `npm` package, before the pre-commit hooks run, it stashes your changes that aren't staged.

This kind of makes sense as it should be running the tests on the code you're committing not all the changes you've made but as a side effect to stashing your unstaged changes, it creates a `patch` and is stored in `/Users/{yourUsername}/.pre-commit/` and I don't necessarily like the idea of accumulating 1000s of patches over the course of the year.

Its better than nothing for now so will leave it in until we find a better solution

**Comment by Danwhy** [comment]
Is there a reason not to use https://www.npmjs.com/package/pre-commit in our Phoenix projects? Do we plan to move away from needing to include node altogether?

**Comment by nelsonic** [comment]
@Danwhy _good_ question. please elaborate. 
were you thinking we can use the NPM `pre-commit` module which still creates the same `git pre-commit hook` and will ensure that our tests run on `localhost` `before` allowing a commit?
That's actually _genius_ given that Phoenix projects `require` node.js by `default` ... let's `try` it!

**Comment by Danwhy** [comment]
Yeah, just installing it with npm and running as we would on a Node project, but with `mix test` instead of tape. 

I've just tried it, and it seems to work fine. We just add our Phoenix test script as the `test` script in the `package.json`, and add that to `pre-commit`. If the tests (or coverage) fail, the commit fails:

![screen shot 2017-03-30 at 17 46 53](https://cloud.githubusercontent.com/assets/8939909/24515971/1442d99c-1571-11e7-9837-68c883e593c7.png)
![screen shot 2017-03-30 at 17 46 46](https://cloud.githubusercontent.com/assets/8939909/24515974/18c653ea-1571-11e7-983b-a6e3ae7c1e68.png)

**Comment by nelsonic** [comment]
@Danwhy you sir, are a _genius_!! 🦄 
Could we add a section to the readme to this effect? (_please/thanks_!)

**Comment by samhstn** [comment]
Using https://www.npmjs.com/package/pre-commit is a great solution for Phoenix projects, but for the Elixir Libraries which we build without Phoenix, we won't have node and this solution doesn't really feel right.

**Comment by nelsonic** [comment]
@shouston3 yeah, for Elixir-only (_Non Phoenix_) projects we do need an Elixir-specific solution ... 
not _urgent_ for the https://github.com/dwyl/amp-elixir project, but a _very_-nice-to-have. 😉

**Comment by nelsonic** [comment]
@shouston3 did you discover how to add the pre-commit hook in Elixir projects _without_ `package.json` ?
there's a "_gap_" on Hex.pm ... https://hex.pm/packages?_utf8=%E2%9C%93&search=pre-commit
![image](https://user-images.githubusercontent.com/194400/30294427-f1915144-9734-11e7-961d-4cc6ac891715.png)

**Comment by samhstn** [comment]
No, I never looked into it, as we were happy with [the solution @Danwhy suggested](https://github.com/dwyl/learn-pre-commit/issues/6#issuecomment-290472082).

But, with Phoenix 1.3 changing it's folder structure so that [node doesn't pollute the root](https://gist.github.com/chrismccord/71ab10d433c98b714b75c886eff17357#move-static-assets-inside-self-contained-assets-directory), the solution may no longer suffice and an elixir pre-commit may be a much better solution.

Do you think it would be good to harness what has already been built in node (or something else), or look at building it entirely in elixir?

**Comment by nelsonic** [comment]
@shouston3 _good_ question ... good point on the Phoenix 1.3 re-org!
I see us writing 20+ "pure" elixir modules in the next year so we _could_ consider _not_ requiring node ...?

_However_ how many people writing code in 2017 _don't_ have Node.js installed on their `localhost` _already_?

The thing I want to _avoid_ is falling into the temptation of using node to run the tests on CI ...

https://travis-ci.org/dwyl/tudo is running the CI tests without touching the `package.json` ...

I'm going to `try` adding a `package.json` file to https://github.com/dwyl/hits-elixir 
and see how "_dirty_" I feel ...

**Comment by ZooeyMiller** [comment]
Haven't looked at it in depth yet, but this may be worth checking out: https://hex.pm/packages/pre_commit_hook

**Comment by ZooeyMiller** [comment]
The one I linked above isn't configurable, so @finnhodgkin and I made a hex module for configurable pre-commit in elixir!

Let us know any thoughts and please try it out! https://github.com/dwyl/elixir-pre-commit
